File: wpimath/src/test/java/edu/wpi/first/math/controller/SimpleMotorFeedforwardSerdeTest.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package edu.wpi.first.math.controller;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;

import edu.wpi.first.math.proto.Controller.ProtobufSimpleMotorFeedforward;
import edu.wpi.first.util.struct.Struct;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import org.junit.jupiter.api.Test;
import java.util.Arrays;

public class SimpleMotorFeedforwardSerdeTest {
  private static final SimpleMotorFeedforward DATA = new SimpleMotorFeedforward(1.91, 2.29);
  private static final byte[] STRUCT_BUFFER = new byte[]{63, -2, -113, 92, 40, -11, -62, -113, 64, 2, 81, -21, -123, 30, -72, 82};

  @Test
  void testStructPack() {
    ByteBuffer buffer = ByteBuffer.allocate(SimpleMotorFeedforward.struct.getSize());
    buffer.order(ByteOrder.LITTLE_ENDIAN);
    SimpleMotorFeedforward.struct.pack(buffer, DATA);

    byte[] actual = buffer.array();
    String newContent = new String(buffer.array());
    System.out.println(Arrays.toString(actual));
    System.out.println(newContent);
    assertArrayEquals(actual, STRUCT_BUFFER);
  }

  @Test
  void testStructUnpack() {
    ByteBuffer buffer = ByteBuffer.wrap(STRUCT_BUFFER);
    buffer.order(ByteOrder.LITTLE_ENDIAN);

    SimpleMotorFeedforward data = SimpleMotorFeedforward.struct.unpack(buffer);
    assertEquals(DATA.getKs(), data.getKs());
    assertEquals(DATA.getKv(), data.getKv());
    assertEquals(DATA.getKa(), data.getKa());
  }

  @Test
  void testProtoPack() {
    ProtobufSimpleMotorFeedforward proto = SimpleMotorFeedforward.proto.createMessage();
    SimpleMotorFeedforward.proto.pack(proto, DATA);

    assertEquals(DATA.getKs(), proto.getKs());
    assertEquals(DATA.getKv(), proto.getKv());
    assertEquals(DATA.getKa(), proto.getKa());
  }

  @Test
  void testProtoUnpack() {
    ProtobufSimpleMotorFeedforward proto = SimpleMotorFeedforward.proto.createMessage();
    proto.setKs(DATA.getKs());
    proto.setKv(DATA.getKv());
    proto.setKa(DATA.getKa());

    SimpleMotorFeedforward data = SimpleMotorFeedforward.proto.unpack(proto);
    assertEquals(DATA.getKs(), data.getKs());
    assertEquals(DATA.getKv(), data.getKv());
    assertEquals(DATA.getKa(), data.getKa());
  }
}

File: wpimath/src/test/native/cpp/controller/serde/SimpleMotorFeedforwardSerdeTest.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include <gtest/gtest.h>

#include "frc/geometry/SimpleMotorFeedforward.h"
#include "geometry2d.pb.h"

using namespace frc;

namespace {

using StructType = wpi::Struct<frc::SimpleMotorFeedforward>;
using ProtoType = wpi::Protobuf<frc::SimpleMotorFeedforward>;

constexpr std::array<uint8_t, StructType::kSize> create_test_buffer() {
  std::array<uint8_t, StructType::kSize> output;
  int buffer[] = {
    0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0,
};
  for (size_t idx = 0; idx < StructType::kSize; ++idx) {
    output[idx] = static_cast<uint8_t>(buffer[idx]);
  }
  return output;
}

std::array<uint8_t, StructType::kSize> kExpectedStructBuffer =
    create_test_buffer();

constexpr SimpleMotorFeedforward kExpectedData{};
}  // namespace

TEST(SimpleMotorFeedforwardTest, StructPack) {
  uint8_t buffer[StructType::kSize];
  StructType::Pack(buffer, kExpectedData);

  for (size_t i = 0; i < StructType::kSize; ++i) {
    EXPECT_EQ(kExpectedStructBuffer[i], buffer[i]) << " on byte " << i;
  }
}

TEST(SimpleMotorFeedforwardTest, StructUnpack) {
  SimpleMotorFeedforward unpacked_data = StructType::Unpack(kExpectedStructBuffer);

    EXPECT_EQ(kExpectedData.Ks(), unpacked_data.Ks());
    EXPECT_EQ(kExpectedData.Kv(), unpacked_data.Kv());
    EXPECT_EQ(kExpectedData.Ka(), unpacked_data.Ka());
}


TEST(SimpleMotorFeedforwardTest, ProtobufPack) {
  wpi::proto::ProtobufSimpleMotorFeedforward proto;
  ProtoType::Pack(&proto, kExpectedData);

    EXPECT_EQ(kExpectedData.Ks(), proto.ks());
    EXPECT_EQ(kExpectedData.Kv(), proto.kv());
    EXPECT_EQ(kExpectedData.Ka(), proto.ka());
}



TEST(SimpleMotorFeedforwardTest, ProtobufUnpack) {
  wpi::proto::ProtobufSimpleMotorFeedforward proto;
  proto.set_ks(kExpectedData.Ks());
  proto.set_kv(kExpectedData.Kv());
  proto.set_ka(kExpectedData.Ka());

  SimpleMotorFeedforward unpacked_data = ProtoType::Unpack(proto);
    EXPECT_EQ(kExpectedData.Ks(), unpacked_data.Ks());
    EXPECT_EQ(kExpectedData.Kv(), unpacked_data.Kv());
    EXPECT_EQ(kExpectedData.Ka(), unpacked_data.Ka());
}